icicle/include/merkle-tree/merkle.cuh

```diff
@@ -0,0 +1,94 @@
+#pragma once
+#ifndef MERKLE_H
+#define MERKLE_H
+
+#include "gpu-utils/device_context.cuh"
+#include "gpu-utils/error_handler.cuh"
+#include "utils/utils.h"
+#include "hash/hash.cuh"
+
+#include <iostream>
+#include <math.h>
+
+using namespace hash;
+
+/**
+ * @namespace merkle_tree
+ * Implementation of the [Merkle tree](https://en.wikipedia.org/wiki/Merkle_tree) builder,
+ * parallelized for the use on GPU
+ */
+namespace merkle_tree {
+  static constexpr size_t GIGA = 1024 * 1024 * 1024;
+
+  /// Bytes per stream
+  static constexpr size_t STREAM_CHUNK_SIZE = 1024 * 1024 * 1024;
+
+  /// Flattens the tree digests and sum them up to get
+  /// the memory needed to contain all the digests
+  static size_t get_digests_len(uint32_t height, uint32_t arity)
+  {
+    size_t digests_len = 0;
+    size_t row_length = 1;
+    for (int i = 0; i < height; i++) {
+      digests_len += row_length;
+      row_length *= arity;
+    }
+
+    return digests_len;
+  }
+
+  /**
+   * @struct TreeBuilderConfig
+   * Struct that encodes various Tree builder parameters.
+   */
+  struct TreeBuilderConfig {
+    device_context::DeviceContext ctx; /**< Details related to the device such as its id and stream id. */
+    int arity;
+    int keep_rows; /**< How many rows of the Merkle tree rows should be written to output. '0' means all of them */
```

Comment:
by rows you mean root is a row, its children are next row and so on?

Reply:
yes